File: src/main/java/org/tindalos/principle/domain/core/CyclesInSubgraph.java
package org.tindalos.principle.domain.core;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.tindalos.principle.domain.core.packages.PackageReference;

/**
 * Tracks cycle detection progress in a package subgraph.
 * Stores cycles grouped by their breaking points and remembers already investigated packages.
 */
public class CyclesInSubgraph {

    public static final int LIMIT = 5;

    private final Set<Package> investigatedPackages = new HashSet<>();
    private final Map<PackageReference, Set<Cycle>> breakingPoints = new HashMap<>();

    public Set<Package> investigatedPackages() {
        return Set.copyOf(investigatedPackages);
    }

    public Map<PackageReference, Set<Cycle>> cycles() {
        return breakingPoints.entrySet().stream()
            .collect(Collectors.toUnmodifiableMap(
                Map.Entry::getKey,
                entry -> Set.copyOf(entry.getValue())));
    }

    public void add(Cycle cycle) {
        boolean exists = breakingPoints.values().stream().anyMatch(cycles -> cycles.contains(cycle));
        if (!exists) {
            breakingPoints.computeIfAbsent(cycle.end(), key -> new HashSet<>()).add(cycle);
        }
    }

    public void rememberPackageAsInvestigated(Package aPackage) {
        investigatedPackages.add(aPackage);
    }

    public void mergeIn(CyclesInSubgraph that) {
        that.breakingPoints.values().forEach(cycles -> cycles.forEach(this::add));
        investigatedPackages.addAll(that.investigatedPackages());
    }

    public Map<PackageReference, Set<Cycle>> mergeBreakingPoints2(Map<PackageReference, Set<Cycle>> breakingPointsInOther) {
        breakingPointsInOther.values().forEach(cycles -> cycles.forEach(this::add));
        return cycles();
    }

    public boolean isBreakingPoint(Package aPackage) {
        Set<Cycle> cyclesForThisBreakingPoint = breakingPoints.get(aPackage.reference());
        return cyclesForThisBreakingPoint != null && cyclesForThisBreakingPoint.size() > LIMIT;
    }

    @Override
    public String toString() {
        return "CyclesInSubgraph [cycles=" + breakingPoints + ", investigatedPackages=" + investigatedPackages() + "]";
    }

    public static CyclesInSubgraph empty() {
        return new CyclesInSubgraph();
    }
}

File: src/main/java/org/tindalos/principle/domain/core/CyclesInSubgraph.scala
package org.tindalos.principle.domain.core

import scala.collection.mutable.Set
import scala.collection.mutable.Map

class CyclesInSubgraph {

  private val investigatedPackages_ = scala.collection.mutable.Set[Package]()
  private val breakingPoints = scala.collection.mutable.Map[PackageReference, scala.collection.mutable.Set[Cycle]]()

  def investigatedPackages = investigatedPackages_.toSet
  
  def cycles: scala.collection.immutable.Map[PackageReference, scala.collection.immutable.Set[Cycle]] = (for ((k, v) <- breakingPoints) yield (k,v.toSet)).toMap

  def add(cycle: Cycle) = breakingPoints.get(cycle.getLast()) match {
    case Some(opt) => opt += cycle
    case None => breakingPoints += cycle.getLast() -> Set[Cycle](cycle)
  }

  def rememberPackageAsInvestigated(aPackage: Package) = investigatedPackages_ += aPackage

  def mergeIn(that: CyclesInSubgraph) = {
    mergeBreakingPoints(that.breakingPoints)
    investigatedPackages_ ++= that.investigatedPackages
  }

  def mergeBreakingPoints(breakingPointsInOther: Map[PackageReference, Set[Cycle]]) =
    for ((k, v) <- breakingPointsInOther) {
      breakingPoints.get(k) match {
        case Some(cyclesForThisBreakingPoint) => cyclesForThisBreakingPoint ++= v
        case None => breakingPoints += k -> v
      }
    }

  def isBreakingPoint(aPackage: Package) = breakingPoints.get(aPackage.reference) match {
    case Some(cyclesForThisBreakingPoint) => cyclesForThisBreakingPoint.size > CyclesInSubgraph.LIMIT
    case None => false
  }

  override def toString() = "CyclesInSubgraph [cycles=" + breakingPoints + ", investigatedPackages=" + investigatedPackages + "]"

}

object CyclesInSubgraph {

  val LIMIT = 25

  def empty() = new CyclesInSubgraph()

}